### PR TITLE
Reimplement cupy.squeeze for increased NumPy compatibility.

### PR DESCRIPTION
**Pull request by takagi**
Rel #1742. This PR reimplements `cupy.squeeze` to increase its compatibility to `numpy.squeeze`.

The latest `cupy.squeeze` has the following incompatibility to `numpy.squeeze`:
- Raises `RuntimeError` when the dimension specified with `axis` is not one, `numpy.squeeze` raises `ValueError`
- Does not work when `axis` is a tuple and is not sorted in ascending order
- Does not work with negative indices

The new implementation of `cupy.squeeze` fairly follows that of `numpy.squeeze`:
- [numpy/core/src/multiarray/methods.c#L213](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/methods.c#L213)
- [numpy/core/src/multiarray/conversion_utils.c#L241](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/conversion_utils.c#L241)
- [numpy/core/src/multiarray/shape.c#L533](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/shape.c#L533)
- [numpy/core/src/multiarray/shape.c#L585](https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/shape.c#L585)

and its test:
- [numpy/core/tests/test_multiarray.py#L1174](https://github.com/numpy/numpy/blob/master/numpy/core/tests/test_multiarray.py#L1174)

The PR's cython code is manually confirmed to produce C++ code with less boxing/unboxing as possible.

### CODE REVIEW

File: tests/cupy_tests/manipulation_tests/test_dims.py [hunk 1]
```diff
@@ -147,13 +147,58 @@ def test_squeeze(self, xp):
...
+    def test_squeze_int_axis1(self, xp):
         a = testing.shaped_arange((1, 2, 1, 3, 1, 1, 4, 1), xp)
         return a.squeeze(axis=2)
 
+    @testing.numpy_cupy_array_equal()
+    def test_squeze_int_axis2(self, xp):
+        a = testing.shaped_arange((1, 2, 1, 3, 1, 1, 4, 1), xp)
+        return a.squeeze(axis=-3)
+
+    def test_squeze_int_axis_failure(self):
```
Comment (unnonouno) [inline tests/cupy_tests/manipulation_tests/test_dims.py]: How about using `numpy_cupy_raises`?

File: tests/cupy_tests/manipulation_tests/test_dims.py [hunk 2]
```diff
@@ -147,13 +147,58 @@ def test_squeeze(self, xp):
...
+            return a.squeeze(axis=(2, 2))
+
+    @testing.numpy_cupy_array_equal()
+    def test_squeeze_scalar1(self, xp):
+        a = testing.shaped_arange((), xp)
+        return a.squeeze(axis=0)
+
+    @testing.numpy_cupy_array_equal()
+    def test_squeeze_scalar2(self, xp):
+        a = testing.shaped_arange((), xp)
```
Comment (unnonouno) [inline tests/cupy_tests/manipulation_tests/test_dims.py]: Add test for n-dim arrays
- shape=() and axis=-2
- shape=() and axis=1

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -517,29 +517,64 @@ cdef class ndarray:
...
+                axis_flags[_axis] = 1
+
+        # Verify that the axes requested are all of size one
+        any_ones = 0
+        for idim in range(ndim):
+            if axis_flags[idim] != 0:
+                if self._shape[idim] == 1:
+                    any_ones = 1
+                else:
+                    raise ValueError('cannot select an axis to squeeze out'
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Insert a space character at end of this str literal.

Comment (unnonouno) [inline cupy/core/core.pyx]: Do we have test for this error?

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -517,29 +517,64 @@ cdef class ndarray:
...
+                                     'which has size not equal to one')
 
-        if vec_axis.size() == 0:
+        # If there were no axes to squeeze out, return the same array
+        if any_ones == 0:
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Do we have a test for this case? I mean `axis=()`.

File: tests/cupy_tests/manipulation_tests/test_dims.py (hunk 2 above)
Comment (takagi) [inline tests/cupy_tests/manipulation_tests/test_dims.py]: Added the cases!

File: cupy/core/core.pyx (hunk 3 above)
Comment (takagi) [inline cupy/core/core.pyx]: I've added the space, the test for the error we have as `test_squeeze_failure` case.
https://github.com/pfnet/chainer/pull/1746/files#diff-85bb3729372a91fb8c4a38547c8c1d91R220

**Comment by takagi** [comment]
I've fixed for the comments!

File: cupy/core/core.pyx (hunk 3 above)
Comment (unnonouno) [inline cupy/core/core.pyx]: OK

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**